### PR TITLE
Fix bugs of buffer allocate

### PR DESCRIPTION
**Pull request by haozech**
1.修正当多个tensor共享同一"cacheread"或"cachewrite" buffer时，bind cuda block/thread导致的buffer大小调整错误问题。
2.添加相关代码生成测试。

### CODE REVIEW

File: cinn/optim/replace_var_with_expr.cc [hunk 1]
```diff
@@ -135,20 +147,23 @@ struct ReplaceVarIndexOfCacheMutator : public ir::IRMutator<> {
 
  private:
   std::map<std::string, ir::Tensor>* global_tensor_map_;
+  std::unordered_set<std::string>* resized_buffer_;
   const Var& var_;
   const Expr& expr_;
   bool blockidx_;
   bool do_replace{false};
+  bool find_replace{false};
```
Comment (Superjomn) [inline cinn/optim/replace_var_with_expr.cc]: member variable's names should have a `_` suffix.
replace to `do_replace_` and `find_replace_`

File: cinn/optim/replace_var_with_expr.cc [hunk 2]
```diff
@@ -135,20 +147,23 @@ struct ReplaceVarIndexOfCacheMutator : public ir::IRMutator<> {
 
  private:
   std::map<std::string, ir::Tensor>* global_tensor_map_;
+  std::unordered_set<std::string>* resized_buffer_;
```
Comment (Superjomn) [inline cinn/optim/replace_var_with_expr.cc]: It's better to use `std::unordered_set<std::string>& resized_buffer_`

File: cinn/optim/replace_var_with_expr.cc (hunk 1 above)
Comment (haozech) [reply cinn/optim/replace_var_with_expr.cc]: Done, thanks

File: cinn/optim/replace_var_with_expr.cc (hunk 2 above)
Comment (haozech) [reply cinn/optim/replace_var_with_expr.cc]: Done, thanks

**Review by Superjomn — APPROVED** [review]
LGTM

**Pull request merged**